cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -2221,6 +2252,67 @@ TEST(Substrait, BasicPlanRoundTrippingEndToEnd) {
   EXPECT_TRUE(expected_table->Equals(*rnd_trp_table));
 }
 
+TEST(Substrait, FilterNamedTable) {
+  compute::ExecContext exec_context;
+  arrow::dataset::internal::Initialize();
+
+  const std::vector<std::string> table_name{"table", "1"};
+  const std::unordered_map<std::string, std::string> metadata = {
+      {Table::kTableNameKey, arrow::internal::JoinStrings(table_name, ".")}};
+  const auto dummy_schema =
+      schema({field("A", int32()), field("B", int32()), field("C", int32())},
+             key_value_metadata(metadata));
+  auto dataset =
+      std::make_shared<dataset::InMemoryDataset>(dummy_schema, RecordBatchVector{});
+
+  auto scan_options = std::make_shared<dataset::ScanOptions>();
+  scan_options->projection = compute::project({}, {});
+  auto filter = compute::equal(compute::field_ref("A"), compute::field_ref("B"));
+
+  auto declarations = compute::Declaration::Sequence(
+      {compute::Declaration(
+           {"scan", dataset::ScanNodeOptions{dataset, scan_options}, "s"}),
+       compute::Declaration({"filter", compute::FilterNodeOptions{filter}, "f"})});
+
+  std::shared_ptr<ExtensionIdRegistry> sp_ext_id_reg = MakeExtensionIdRegistry();
+  ExtensionIdRegistry* ext_id_reg = sp_ext_id_reg.get();
+  ExtensionSet ext_set(ext_id_reg);
```

Comment:
You probably don't need to create an `ExtensionIdRegistry` (it will just use the default if you do not supply one).